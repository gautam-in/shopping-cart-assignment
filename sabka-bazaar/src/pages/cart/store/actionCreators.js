import { ADD_TO_CART } from "./types";

export const addToCart = (product) => {
  return {
    type: ADD_TO_CART,
    payload: product,
  };
};
