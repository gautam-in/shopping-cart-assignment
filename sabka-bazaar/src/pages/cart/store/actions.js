import { ADD_TO_CART, UPDATE_CART } from "./types";
import { createAction } from "@reduxjs/toolkit";

export const addToCart = createAction(ADD_TO_CART);

export const updateCartItems = createAction(UPDATE_CART);
