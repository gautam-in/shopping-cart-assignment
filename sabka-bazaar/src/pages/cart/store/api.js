export const postCartItemToServer = async (options) => {
  try {
    const response = await fetch(
      `${process.env.REACT_APP_BASEURL}/addToCart`,
      options
    );
    return response.json();
  } catch (error) {
    throw new Error(error.message);
  }
};
