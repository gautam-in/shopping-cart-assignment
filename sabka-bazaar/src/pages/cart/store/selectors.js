import { createSelector } from "@reduxjs/toolkit";

const selectSlice = (state) => state["cart"] || {};

export const selectIsAddingToCartLoading = createSelector(
  [selectSlice],
  (state) => state.addToCartLoading
);

export const selectCartItems = createSelector(
  [selectSlice],
  (state) => state.cart
);

export const selectCartItemsCount = createSelector(
  [selectCartItems],
  (cartItems) =>
    cartItems.reduce(
      (cartItemsCount, cartItem) => cartItemsCount + cartItem.quantity,
      0
    )
);

export const selectCartTotal = createSelector([selectCartItems], (cartItems) =>
  cartItems.reduce(
    (cartTotal, cartItem) => cartTotal + cartItem.quantity * cartItem.price,
    0
  )
);
