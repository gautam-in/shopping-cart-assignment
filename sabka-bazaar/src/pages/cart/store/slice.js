import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  addToCartLoading: true,
  error: "",
  cart: [],
};

export const cartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    addToCartLoading: (state, action) => {
      state.addToCartLoading = action.payload;
    },
    addToCartSuccess: (state, action) => {
      state.cart = action.payload;
    },
    addToCartError: (state, action) => {
      state.error = action.payload;
    },
    updateCart: (state, action) => {
      state.cart = action.payload;
    },
  },
});

export const {
  addToCartLoading,
  addToCartSuccess,
  addToCartError,
  updateCart,
} = cartSlice.actions;
export default cartSlice.reducer;
