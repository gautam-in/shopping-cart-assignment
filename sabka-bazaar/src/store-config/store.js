import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";
import homeReducer from "../pages/home/store/slice";
import productsReducer from "../pages/products/store/slice";
import cartReducer from "../pages/cart/store/slice";
import rootSaga from "./rootSaga";
const sagaMiddleware = createSagaMiddleware();
export const store = configureStore({
  reducer: {
    home: homeReducer,
    products: productsReducer,
    cart: cartReducer,
  },
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);
