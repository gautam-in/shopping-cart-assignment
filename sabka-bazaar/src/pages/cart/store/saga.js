import { ADD_TO_CART, UPDATE_CART } from "./types";
import {
  addToCartLoading,
  addToCartSuccess,
  addToCartError,
  updateCart,
} from "./slice";
import { call, put, takeLatest } from "redux-saga/effects";
import { postCartItemToServer } from "./api";

const addToCartHelper = ({ cartItems, productToAdd }) => {
  const existingCartItem = cartItems.find((cartItem) => {
    return cartItem.id === productToAdd.id;
  });
  if (existingCartItem) {
    return cartItems.map((cartItem) =>
      cartItem.id === productToAdd.id
        ? { ...cartItem, quantity: cartItem.quantity + 1 }
        : cartItem
    );
  }

  return [...cartItems, { ...productToAdd, quantity: 1 }];
};

function* handleCartData({ payload }) {
  const options = {
    method: "POST",
    data: { id: payload.productToAdd.id },
  };

  try {
    const response = yield call(postCartItemToServer, options);
    if (response) {
      const cartItems = addToCartHelper(payload);
      yield put(addToCartSuccess(cartItems));
    }
  } catch (error) {
    yield put(addToCartError(error.message));
  } finally {
    yield put(addToCartLoading(false));
  }
}

function* updateCartHandler({ payload }) {
  yield put(updateCart(payload));
}

function* cartSaga() {
  yield takeLatest(ADD_TO_CART, handleCartData);
  yield takeLatest(UPDATE_CART, updateCartHandler);
}

export default cartSaga;
