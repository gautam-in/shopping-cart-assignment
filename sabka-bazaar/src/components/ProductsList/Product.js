import { Button } from "../Button/Button";
import PropTypes from "prop-types";
import { useDispatch, useSelector } from "react-redux";
import { useCallback } from "react";
import { addToCart } from "../../pages/cart/store/actions";
import React from "react";
import { selectCartItems } from "../../pages/cart/store/selectors";
import { ButtonWithPrice } from "../Button/ButtonWithPrice";
const Product = ({ name, id, imageURL, description, price }) => {
  const dispatch = useDispatch();
  const cartItems = useSelector(selectCartItems);
  const addToCartHandler = useCallback(
    ({ name, id, imageURL, description, price }) => {
      dispatch(
        addToCart({
          cartItems,
          productToAdd: { name, id, imageURL, description, price },
        })
      );
    },
    [{ name, id, imageURL, description, price }]
  );
  return (
    <div key={id} className="product-container">
      <div className="product-title">{name}</div>
      <div className="product-img-container">
        <img
          src={imageURL}
          alt={name}
          className="product-img"
          width="184"
          height="184"
        />
      </div>

      <div className="product-desc">
        <p>{description}</p>
      </div>
      <div className="buy-now">
        <p>Mrp Rs. {price}</p>
        <div className="button-container">
          <Button
            title="Buy Now"
            price={price}
            onClick={() =>
              addToCartHandler({ name, id, imageURL, description, price })
            }
          />
          <ButtonWithPrice
            title="Buy Now"
            price={price}
            onClick={() =>
              addToCartHandler({ name, id, imageURL, description, price })
            }
          />
        </div>
      </div>
    </div>
  );
};
Product.propTypes = {
  name: PropTypes.string,
  id: PropTypes.string,
  imageURL: PropTypes.string,
  description: PropTypes.string,
  price: PropTypes.number,
};

export default React.memo(Product);
