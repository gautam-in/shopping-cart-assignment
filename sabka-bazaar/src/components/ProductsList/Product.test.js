import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import Product from "./Product";
import { store } from "../../store-config/store";

describe("Product component", () => {
  it("should match snapshot", () => {
    let mockProps = {
      name: "",
      id: "",
      imageURL: "",
      description: "",
      price: 0,
    };

    const container = render(
      <Provider store={store}>
        <Product props={mockProps} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
