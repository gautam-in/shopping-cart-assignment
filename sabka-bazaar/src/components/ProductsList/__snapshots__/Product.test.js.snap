// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="product-container"
      >
        <div
          class="product-title"
        />
        <div
          class="product-img-container"
        >
          <img
            class="product-img"
            height="184"
            width="184"
          />
        </div>
        <div
          class="product-desc"
        >
          <p />
        </div>
        <div
          class="buy-now"
        >
          <p>
            Mrp Rs. 
          </p>
          <div
            class="button-container"
          >
            <button
              class="button"
            >
              <div>
                Buy Now
              </div>
            </button>
            <button
              class="button-with-price"
            >
              Buy Now
               @ Rs.
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="product-container"
    >
      <div
        class="product-title"
      />
      <div
        class="product-img-container"
      >
        <img
          class="product-img"
          height="184"
          width="184"
        />
      </div>
      <div
        class="product-desc"
      >
        <p />
      </div>
      <div
        class="buy-now"
      >
        <p>
          Mrp Rs. 
        </p>
        <div
          class="button-container"
        >
          <button
            class="button"
          >
            <div>
              Buy Now
            </div>
          </button>
          <button
            class="button-with-price"
          >
            Buy Now
             @ Rs.
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
